test/Rings/FreeAssociativeAlgebraIdeal.jl

```diff
@@ -122,3 +122,11 @@ end
     @test is_groebner_basis(gb_bb)
     @test S1 == S1_alt
 end
+
+@testset "FreeAssociativeAlgebraIdeal.groebner_basis.f4" begin
+    R, u = free_associative_algebra(QQ,[:u])
+    v = FreeAssociativeAlgebraElem{QQFieldElem}[u[1]-one(R)]
```

Comment:
So this intentionally creates the problem by using the wrong type (of course, so you can test your workaround). If this code instead did
```suggestion
    v = [u[1]-one(R)]
```
or
```suggestion
    v = AbstractAlgebra.Generic.FreeAssociativeAlgebraElem{QQFieldElem}[u[1]-one(R)]
```
then all would be fine.

This leads me to ask: where do you run into this "in nature"?

Is the problem that some code accidentally generated such a vector but was not caught (i.e. stricter input validation is needed) ?

Or do you really "need" such vectors (but why?)

In the former case, I'd simply make the IdealGens constructors more strict. In the latter case they'd also need to learn to convert such inputs.

Reply:
Thanks for looking into this. I would say this happens in nature due to the mistake of believing that the element type is 'FreeAssociativeAlgebraElem' rather than 'Generic.FreeAssociativeAlgebraElem'. I haven't found Oscar code that produces a vector like that.


Nevertheless I stumbled upon this issue so many times over the last few years that I thought I should try to fix it.

I don't know the technical reasons why the concrete type has to be called 'Generic.FreeAssociativeAlgebraElem' whereas the generic type is called 'FreeAssociativeAlgebraElem'. I thought that, at least, no 'normal' user (myself included) has to be exposed to that. This is my fix to allow for the intuitive input.

However, your suggestion to fix the Idealgens is fine by me. I just don't know where to implement that change.

Reply:
> Thanks for looking into this. I would say this happens in nature due to the mistake of believing that the element type is 'FreeAssociativeAlgebraElem' rather than 'Generic.FreeAssociativeAlgebraElem'. I haven't found Oscar code that produces a vector like that.

OK, that's somewhat reassuring!

> Nevertheless I stumbled upon this issue so many times over the last few years that I thought I should try to fix it.

It's good to fix such a persistent annoyance, thank you for making the effort (and my PR is not a replacement for this one, more a complement).

That side, I wonder how exactly you stumble over the issue? By declaring a `Vector{T}` with `T` accidentally of the wrong type is one way. Are there other scenarios?

(To be clear: this is not me doubting you, this is my trying to get a better grasp of the problem which then might inspire additional ideas for how things could be improved)

> 
> I don't know the technical reasons why the concrete type has to be called 'Generic.FreeAssociativeAlgebraElem' whereas the generic type is called 'FreeAssociativeAlgebraElem'.

Well... nothing "has" to be called some way, we could also call them, say `FooType` and `BarType`. But in the end, it's no different for, say, matrices: we have the generic type `MatElem` (or rather, `MatElem{T}`), and then subtypes such as `Generic.MatSpaceElm{T}` or `ZZMatrix`. The only difference in the context of free associative algebras is that currently we have only have one concrete type, and unfortunately it has the exact same name as the abstract type; so to distinguish them, one has to prefix one with the name of the module it's been declared in (`Generic`).


> I thought that, at least, no 'normal' user (myself included) has to be exposed to that. This is my fix to allow for the intuitive input.

Actually, I'd indeed expect a "normal user" to not be exposed to that (other than by seeing it printed in some places). A "normal user" in my mind would normally not use the type at all.  But I guess the term "normal user" is simply not well-defined ;-). What constitutes "normal" and also what constitutes a "user" can vary a lot :-).

Hence my question from above: how are you using the code that leads to this being such a problem?

> 
> However, your suggestion to fix the Idealgens is fine by me. I just don't know where to implement that change.

Sure, no worries, that's why I offered to do it -- see PR #5525 for the result

Reply:
So this was the code that motivated me to write that change. I created a vector with the wrong type, forgot about it, and later encountered weird issues when trying to create the ideal.

```
function create_tensor_ideal(gb::Vector{Generic.FreeAssociativeAlgebraElem{T}}) where {T}
  R_old = parent(gb[1])
  N = Int(sqrt(length(gens(R_old))))
  N2 = N * N
  R, u, v  = free_associative_algebra(QQ, :u=>(1:N,1:N), :v=>(1:N,1:N))
  ret = Vector{FreeAssociativeAlgebraElem{T}}()
  sizehint!(ret, length(gb)*length(gb) + N2*N2)
  for f in gb 
      push!(ret, R(f.coeffs, f.exps))
      push!(ret, R(f.coeffs, increase_by(f.exps,N2))) 
  end
  for i in 1:N2
      for j in 1:N2
        push!(ret, R[i]*R[N2+j] - R[N2+j]*R[i])
      end
  end

  return ret, u, v
end
```
Specifically the line
```
 ret = Vector{FreeAssociativeAlgebraElem{T}}()
```

That's where I made the mistake before. Simply by forgetting about it. Maybe I'm just bad at programming, but I can only imagine how frustrating it is for someone who doesn't even know the difference between the two versions of the type.

> That side, I wonder how exactly you stumble over the issue? By declaring a `Vector{T}` with `T` accidentally of the wrong type is one way. Are there other scenarios?
> 
exactly this.

> Well... nothing "has" to be called some way, we could also call them, say `FooType` and `BarType`. But in the end, it's no different for, say, matrices: we have the generic type `MatElem` (or rather, `MatElem{T}`), and then subtypes such as `Generic.MatSpaceElm{T}` or `ZZMatrix`. The only difference in the context of free associative algebras is that currently we have only have one concrete type, and unfortunately it has the exact same name as the abstract type; so to distinguish them, one has to prefix one with the name of the module it's been declared in (`Generic`).

Makes sense. I just wanted to say that I don't know if this has to be exposed to the user.

> 
> > I thought that, at least, no 'normal' user (myself included) has to be exposed to that. This is my fix to allow for the intuitive input.
> 
> Actually, I'd indeed expect a "normal user" to not be exposed to that (other than by seeing it printed in some places). A "normal user" in my mind would normally not use the type at all. But I guess the term "normal user" is simply not well-defined ;-). What constitutes "normal" and also what constitutes a "user" can vary a lot :-).

Gotcha xD
I am just saying that if I have understood the concept of Oscar correctly, then it is trying to make it as easy as possible to start using it. However, the issue at hand seems like a roadblock many would encounter.

Reply:
> I am just saying that if I have understood the concept of Oscar correctly, then it is trying to make it as easy as possible to start using it.

Short version: trying to make it "easy to use" is certainly something we aspire to, but of course as with any aspiration, we sometimes fail -- esp. since it is in tension with other aspirations and you have to make compromises.

In this case, I think it's a mix between OSCAR needs to be better (your PR here and my other one work towards this), and things you did that IMHO are *not* what a "typical user" would do, but what someone who is developing library code learned to do ("avoid type instability" -- good on you! but it's technical and difficult) mixed with some bad luck due to confusing type names.

That's not meant as a slight against you at all, rather it is an unfortunate consequence of the fact that OSCAR is big and complex, combined with some unfortunate pitfalls in Julia. We just have to learn to live with these things and how to deal with them.

In this case, I would strongly suggest to avoid patterns like
```julia
  R, u, v  = free_associative_algebra(QQ, :u=>(1:N,1:N), :v=>(1:N,1:N))
  ret = Vector{FreeAssociativeAlgebraElem{T}}()
```
and instead would always write
```julia
  R, u, v  = free_associative_algebra(QQ, :u=>(1:N,1:N), :v=>(1:N,1:N))
  ret = Vector{elem_type(R)}()
```
or rather
```julia
  R, u, v  = free_associative_algebra(QQ, :u=>(1:N,1:N), :v=>(1:N,1:N))
  ret = elem_type(R)()
```
Not only does this avoid the problem you run into, but it also has the advantage that you are not hardcoding assumptions about the implementation of the ring returned by `free_associative_algebra` anymore. And it is even shorter, and makes the intent clearer ("this list is meant to store elements of `R`").


---

By the way, I've now revised this PR to take advantage of some generic code for ideals in AA: there is code there which already takes care of dealing with generator `Vector` objects of the "wrong" type. By narrowing the signature of your `ideal` method, that general code has a chance to kick in. I also removed a second redundant `ideal` method.

To be clear, I do not think you had a chance of knowing about these methods: they are not documented anywhere! In fact so far there is no specification for how to implement ideals in Oscar or AbstractAlgebra! To improve this, I just merged a PR to AA which adds a first version of such a specification, and hope to soon be able to extend it and add more generic code to AA. That will in fact allow remove some more code here, and also for other ideal implementations in OSCAR.